File: plugins/cli-with-claude-skill/skills/cli-with-claude-skill/template/go/skill_install.go
// Drop this file into the cobra-based main package of your CLI.
//
// Wire-up checklist (in your main.go init / rootCmd setup):
//
//   rootCmd.AddCommand(installClaudeSkillCmd)
//
//   rootCmd.PersistentPreRun = func(cmd *cobra.Command, args []string) {
//       // ...your existing PersistentPreRun work...
//       checkSkillNotice()
//   }
//
// Replace `mytool` with your actual binary name and `<your-module>` with the
// Go module path that hosts internal/skill.
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/spf13/cobra"

	"<your-module>/internal/skill"
)

const toolName = "mytool"

var embeddedSkill = skill.Embedded

func claudeSkillPath() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".claude", "skills", toolName, "SKILL.md")
}

func parseSkillVersion(content string) string {
	for _, block := range strings.SplitN(content, "---", 3) {
		for _, l := range strings.Split(block, "\n") {
			l = strings.TrimSpace(l)
			if strings.HasPrefix(l, "version:") {
				return strings.TrimSpace(strings.TrimPrefix(l, "version:"))
			}
		}
	}
	return ""
}

func parseSemver(v string) ([3]int, bool) {
	var out [3]int
	v = strings.TrimPrefix(strings.TrimSpace(v), "v")
	if i := strings.IndexAny(v, "-+"); i >= 0 {
		v = v[:i]
	}
	parts := strings.Split(v, ".")
	if len(parts) == 0 || len(parts) > 3 {
		return out, false
	}
	for i, p := range parts {
		n, err := strconv.Atoi(p)
		if err != nil || n < 0 {
			return out, false
		}
		out[i] = n
	}
	return out, true
}

func shouldShowSkillOutdatedNotice(installed, embedded string) bool {
	if embedded == "" {
		return false
	}
	if installed == "" {
		return true
	}
	inst, instOK := parseSemver(installed)
	emb, embOK := parseSemver(embedded)
	if !instOK || !embOK {
		return installed != embedded
	}
	for i := 0; i < 3; i++ {
		if emb[i] != inst[i] {
			return emb[i] > inst[i]
		}
	}
	return false
}

func checkSkillNotice() {
	installedData, err := os.ReadFile(claudeSkillPath())
	if err != nil {
		fmt.Fprintf(os.Stderr, "Tip: Run `%s install-claude-skill` to add Claude Code skill support.\n", toolName)
		return
	}
	installedVersion := parseSkillVersion(string(installedData))
	embeddedVersion := parseSkillVersion(embeddedSkill)
	if shouldShowSkillOutdatedNotice(installedVersion, embeddedVersion) {
		fmt.Fprintln(os.Stderr, "Notice: Claude skill is outdated. Auto-updating...")
		if err := cmdInstallClaudeSkill(); err != nil {
			fmt.Fprintf(os.Stderr, "  failed: %v. Run `%s install-claude-skill` to update manually.\n", err, toolName)
		}
	}
}

func cmdInstallClaudeSkill() error {
	path := claudeSkillPath()
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return fmt.Errorf("creating skill directory: %w", err)
	}
	if err := os.WriteFile(path, []byte(embeddedSkill), 0644); err != nil {
		return fmt.Errorf("writing skill file: %w", err)
	}
	fmt.Printf("Claude Code skill installed at %s\n", path)
	return nil
}

var installClaudeSkillCmd = &cobra.Command{
	Use:   "install-claude-skill",
	Short: "Install/update the Claude Code skill into ~/.claude/skills/" + toolName + "/",
	RunE: func(cmd *cobra.Command, args []string) error {
		return cmdInstallClaudeSkill()
	},
}

File: npm/cli-with-claude-skill/skill/template/node/skillInstall.ts
// Drop into your commander-based CLI.
//
// Project layout:
//   src/skill.md          (the embedded skill markdown)
//   src/skillInstall.ts   (this file)
//   src/cli.ts            (your commander entrypoint)
//
// Wire-up in cli.ts:
//
//   import { Command } from "commander";
//   import { registerSkillCommand, checkSkillNotice } from "./skillInstall.js";
//
//   const program = new Command();
//   program.name("mytool");
//   registerSkillCommand(program);
//   program.hook("preAction", () => checkSkillNotice());
//   // ...your other subcommands
//   program.parseAsync(process.argv);
//
// Replace `mytool` with your actual binary name.
//
// Bundling note: this file uses `fs.readFileSync` at module load to inline
// `skill.md`. If you ship via tsup/esbuild/etc, configure the bundler to
// copy `skill.md` next to the bundle, or switch to a plugin that inlines
// .md files at build time and assign EMBEDDED accordingly.

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import type { Command } from "commander";

export const TOOL_NAME = "mytool";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const EMBEDDED: string = fs.readFileSync(path.join(__dirname, "skill.md"), "utf8");

export function skillPath(): string {
  return path.join(os.homedir(), ".claude", "skills", TOOL_NAME, "SKILL.md");
}

export function parseVersion(content: string): string | null {
  const blocks = content.split("---");
  for (let i = 0; i < Math.min(blocks.length, 3); i++) {
    for (const line of blocks[i].split("\n")) {
      const t = line.trim();
      if (t.startsWith("version:")) {
        return t.slice("version:".length).trim();
      }
    }
  }
  return null;
}

export function parseSemver(v: string): [number, number, number] | null {
  let s = v.trim();
  if (s.startsWith("v")) s = s.slice(1);
  s = s.split(/[-+]/)[0];
  const parts = s.split(".");
  if (parts.length === 0 || parts.length > 3) return null;
  const out: [number, number, number] = [0, 0, 0];
  for (let i = 0; i < parts.length; i++) {
    const n = Number(parts[i]);
    if (!Number.isInteger(n) || n < 0) return null;
    out[i] = n;
  }
  return out;
}

export function shouldShowOutdated(installed: string | null, embedded: string | null): boolean {
  if (!embedded) return false;
  if (!installed) return true;
  const a = parseSemver(installed);
  const b = parseSemver(embedded);
  if (!a || !b) return installed !== embedded;
  for (let i = 0; i < 3; i++) {
    if (a[i] !== b[i]) return b[i] > a[i];
  }
  return false;
}

export async function installClaudeSkill(): Promise<void> {
  const p = skillPath();
  await fs.promises.mkdir(path.dirname(p), { recursive: true, mode: 0o700 });
  await fs.promises.writeFile(p, EMBEDDED, { mode: 0o644 });
  console.log(`Claude Code skill installed at ${p}`);
}

export function checkSkillNotice(): void {
  let installed: string;
  try {
    installed = fs.readFileSync(skillPath(), "utf8");
  } catch {
    console.error(`Tip: Run \`${TOOL_NAME} install-claude-skill\` to add Claude Code skill support.`);
    return;
  }
  const installedV = parseVersion(installed);
  const embeddedV = parseVersion(EMBEDDED);
  if (shouldShowOutdated(installedV, embeddedV)) {
    console.error("Notice: Claude skill is outdated. Auto-updating...");
    installClaudeSkill().catch((e) => {
      console.error(`  failed: ${e}. Run \`${TOOL_NAME} install-claude-skill\` to update manually.`);
    });
  }
}

export function registerSkillCommand(program: Command): void {
  program
    .command("install-claude-skill")
    .description(`Install/update the Claude Code skill into ~/.claude/skills/${TOOL_NAME}/`)
    .action(async () => {
      await installClaudeSkill();
    });
}
